packages/block-editor/src/components/block-drop-zone/index.js

```diff
@@ -109,18 +107,48 @@ export default function useBlockDropZone( { element, clientId, rootClientId } )
 		}
 
 		const dstIndex = clientId ? getBlockIndex( clientId, rootClientId ) : undefined;
-		const positionIndex = getInsertIndex( position );
+		const positionIndex = blockIndex;
 		// If the block is kept at the same level and moved downwards,
 		// subtract to account for blocks shifting upward to occupy its old position.
 		const insertIndex = dstIndex && srcIndex < dstIndex && isSameLevel( srcRootClientId, rootClientId ) ? positionIndex - 1 : positionIndex;
 		moveBlockToPosition( srcClientId, srcRootClientId, rootClientId, insertIndex );
 	}
 
-	return useDropZone( {
+	const { position } = useDropZone( {
 		element,
 		onFilesDrop,
 		onHTMLDrop,
 		onDrop,
 		isDisabled: isLockedAll,
+		withExactPosition: true,
+	} );
+
+	useEffect( () => {
+		if ( position ) {
+			const { y } = position;
+			const rect = element.current.getBoundingClientRect();
+
+			const offset = y - rect.top;
+			const target = Array.from( element.current.children ).find( ( blockEl ) => {
+				return blockEl.offsetTop + ( blockEl.offsetHeight / 2 ) > offset;
+			} );
```

Comment:
Is it possible to take into consideration the blocks that are laid out horizontally (I'm not sure it was the case before)

Reply:
It's not possible in `master` right now, so I'd prefer to tackle this in a separate PR to keep this small and contained.